cuda/preconditioner/jacobi_kernels.cu

```diff
@@ -131,39 +181,52 @@ __launch_bounds__(warps_per_block *cuda_config::warp_size) adaptive_generate(
                               : 0;
     auto perm = subwarp.thread_rank();
     auto trans_perm = subwarp.thread_rank();
-    auto prec = precision_reduction(2, 2);
+    auto prec_descriptor = uint32{} - 1;
```

Comment:
[Value initialization](https://en.cppreference.com/w/cpp/language/value_initialization) with an empty initialization list [zero-initializes](https://en.cppreference.com/w/cpp/language/zero_initialization) non-class types.
Since both types `unsigned int` and `int` have the same conversion rank, the [signed type is cast to the unsigned one](https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions), and the result is `uint32`.

Finally since unsigned integers implement modular arithmetic, the result will in this case be `0xFFFFFFFF`.

I don't think that `uint32{}` is a very strange syntax, since it's usually used all the time to call the default constructor / zero initialize data (e.g. `template <typename T> f(T t = T{});`).

The mix between the signed / unsigned part is actually no ideal, but since `uint32` is of fixed width, the cleanest implementation would probably be:

```c++
auto prec_descriptor = uint32{0xffffffffu};
```

Reply:
Good to know, I was worried that it would be somewhat similar to `int temp;`, so no / a random initialization.
I would prefer your last suggestion or something like:
```C++
auto prec_descriptor = ~uint32{0};        // or without the 0 in it
```
to make sure the type is clear and every bit is set.

Reply:
`~uint32{}` is actually the most accurate choice. Thanks, that possibility didn't even cross my mind for some reason...